db/db_impl.cc

```diff
@@ -3188,55 +3226,114 @@ Status DBImpl::IngestExternalFile(
+            auto* cfd =
+                static_cast<ColumnFamilyHandleImpl*>(args[i].column_family)
+                    ->cfd();
+            status = FlushMemTable(cfd, flush_opts,
+                                   FlushReason::kExternalFileIngestion,
+                                   true /* writes_stopped */);
+            mutex_.Lock();
+            if (!status.ok()) {
+              break;
+            }
+          }
+        }
       }
     }
-
-    // Run the ingestion job
+    // Run ingestion jobs.
+    // After each run, versions_->last_seqno will be changed.
+    const SequenceNumber last_seqno = versions_->LastSequence();
     if (status.ok()) {
-      status = ingestion_job.Run();
+      for (size_t i = 0; i != num_cfs; ++i) {
+        status = ingestion_jobs[i].Run();
+        if (!status.ok()) {
+          break;
+        }
+      }
     }
-
-    // Install job edit [Mutex will be unlocked here]
-    auto mutable_cf_options = cfd->GetLatestMutableCFOptions();
     if (status.ok()) {
+      autovector<ColumnFamilyData*> cfds_to_commit;
+      autovector<const MutableCFOptions*> mutable_cf_options_list;
+      autovector<autovector<VersionEdit*>> edit_lists;
+      uint32_t num_entries = 0;
+      for (size_t i = 0; i != num_cfs; ++i) {
+        auto* cfd =
+            static_cast<ColumnFamilyHandleImpl*>(args[i].column_family)->cfd();
+        if (cfd->IsDropped()) {
+          continue;
+        }
+        cfds_to_commit.push_back(cfd);
+        mutable_cf_options_list.push_back(cfd->GetLatestMutableCFOptions());
+        autovector<VersionEdit*> edit_list;
+        edit_list.push_back(ingestion_jobs[i].edit());
+        edit_lists.push_back(edit_list);
+        ++num_entries;
+      }
+      // Mark the version edits as an atomic group
+      for (auto& edits : edit_lists) {
+        assert(edits.size() == 1);
+        edits[0]->MarkAtomicGroup(--num_entries);
+      }
+      assert(0 == num_entries);
       status =
-          versions_->LogAndApply(cfd, *mutable_cf_options, ingestion_job.edit(),
-                                 &mutex_, directories_.GetDbDir());
+          versions_->LogAndApply(cfds_to_commit, mutable_cf_options_list,
+                                 edit_lists, &mutex_, directories_.GetDbDir());
     }
-    if (status.ok()) {
-      InstallSuperVersionAndScheduleWork(cfd, &sv_context, *mutable_cf_options);
+    // If writing to MANIFEST fails, then restore last sequence number
+    if (!status.ok()) {
+      versions_->SetLastAllocatedSequence(last_seqno);
+      versions_->SetLastPublishedSequence(last_seqno);
+      versions_->SetLastSequence(last_seqno);
+    } else {
+      for (size_t i = 0; i != num_cfs; ++i) {
+        auto* cfd =
+            static_cast<ColumnFamilyHandleImpl*>(args[i].column_family)->cfd();
+        if (!cfd->IsDropped()) {
+          InstallSuperVersionAndScheduleWork(cfd, &sv_ctxs[i],
+                                             *cfd->GetLatestMutableCFOptions());
```

Comment:
Does it mean that it's not strictly atomic? Data ingested to some CFs can be visible before others? If MyRocks is OK about it, it's OK, since it's atomic from recovery's point of the view but we need to clearly document the expectation to users.

Reply:
>> Does it mean that it's not strictly atomic? Data ingested to some CFs can be visible before others? If MyRocks is OK about it, it's OK, since it's atomic from recovery's point of the view but we need to clearly document the expectation to users.

This is not good from consistency point of view as user may observe a in-between state. Of course, that is assuming what @siying saying is indeed the case. If this is atomic then awesome (I haven't got a chance to read the code yet). Though to be perfectly honest, in our scenarios it is probably OK as atomic ingestion API tends to be called for brand new tables where there are no access, so it is still better than what we have now. Still, I'd prefer this to be atomic.

Reply:
@siying I'm afraid you are right here. I should have made it clearer. Users can see in-between states in which some column families have ingested data, while others do not.
@yizhang82 , `Get` does not try to acquire db mutex. Therefore, we cannot use lock-based approach. It's also difficult to do it with atomic instructions because it involves updating multiple 8-byte variables on different cache lines.

Reply:
@riversand963 Are you talking about DB::Get? What if we use snapshots or Transaction::Get? Are those cases OK (in the sense that it won't observe inconsistent state)?

Reply:
I did some investigation and correct me if I am wrong. @yizhang82 @siying 
Whether you use `DB::Get` or `Transaction::Get`, you will see the latest state if you do not specify a snapshot in `ReadOptions`. Even if you take a snapshot, and later read from this snapshot, it is still possible that you see data ingested when you use the `ingest_behind` option because this option assigns a `0` as its sequence number indicating the data started to exist in the past.

Reply:
We don't use `ingest_behind` in MyRocks.

Reply:
Then taking a snapshot will prevent the application from seeing data that is being ingested concurrently because the ingested data will have the latest sequence number larger than the snapshot.

Reply:
@yizhang82 , does this work for you?

Reply:
@riversand963 thanks. I'd like chat about this offline to clarify a few things. We can come back and update the thread later.

Reply:
@yizhang82 , to clarify, `InstallSuperVersionAndScheduleWork` does not release the db mutex, thus installing the superversions for multiple column families is inside a critical section. Point lookups (`Get`) temporarily locks the DB mutex for a short period of time.
I've added a test in the latest commit showing how to avoid observing mixed state with snapshot. For the rest, we can chat offline.

Reply:
Chatted with @siying , we should let the ingested files have the same seqno. Pushed a new commit.